File: JNA/Java/tests/hash/hash.java
public class hash
{
   public static void main(String[] args) {
      ArcanaJNI aJNI = new ArcanaJNI();

      aJNI.ktb_init();

      String hello = "Hello1";
      byte[] helloB = hello.getBytes();

      ArcanaJNI.ktb_hash_algo_t HASH_ALGO = ArcanaJNI.ktb_hash_algo_t.KTB_HASH_ALGO_SHA512;

      long hash_len = aJNI.ktb_hash_get_len_from_algo(HASH_ALGO);
      System.out.println("hash_len : " + hash_len);

      byte[] hash = new byte[(int) (long)hash_len];

      aJNI.ktb_hash_block(HASH_ALGO, helloB, hash);

      StringBuilder sb = new StringBuilder();
      for(byte b : hash) 
         sb.append(String.format("%02X ", b));
      System.out.println(sb.toString());

      aJNI.ktb_clear();
   }
}

File: JNA/Java/tests/hash/performance test/hash.c
#include <stdio.h>
#include <time.h>
#include <stdint.h>
#include <ktb.h>
#include <ktb/hash.h>

int hashString() {
	char *hello = "Hello1";
	size_t hash_len = ktb_hash_get_len_from_algo(KTB_HASH_ALGO_SHA512);
   printf("hash_len : %ld\n", hash_len);
	uint8_t *hash = malloc(hash_len);

	if(ktb_hash_block(KTB_HASH_ALGO_SHA512, hello, strlen(hello), hash, hash_len) == KTB_ERR_NO_ERROR)
		printf("ktb_hash_block() ok\n");

   int i;
   for(i = 0; i < hash_len; ++i)
      printf("%02X ", hash[i]);
   printf("\n");
   hello = NULL;
   for(i = 0; i < hash_len; ++i)
      hash[i] = 0;
   free(hash);

	return 1;
}

int main(void) {

	struct timespec tsi, tsf;

	clock_gettime(CLOCK_MONOTONIC, &tsi);
	if(ktb_init() == KTB_ERR_NO_ERROR) 
		printf("ktb_init() ok\n");

	if(!hashString()) {
		ktb_clear();
		exit(1);
	}

	ktb_clear();

	clock_gettime(CLOCK_MONOTONIC, &tsf);
	double elaps_s = difftime(tsf.tv_sec, tsi.tv_sec);
	long elaps_ns = tsf.tv_nsec - tsi.tv_nsec;
	printf("Elapsed time : %lf ms\n", elaps_s + ((double)elaps_ns) / 1.0e6);
	return 0;
}


